discovery/kubernetes/kubernetes.go

```diff
@@ -229,27 +235,39 @@ func (d *Discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
 			e := d.client.CoreV1().Endpoints(namespace)
 			elw := &cache.ListWatch{
 				ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+					options.FieldSelector = d.fieldSelector
+					options.LabelSelector = d.labelSelector
```

Comment:
What happens if the endpoints match, but not the associated services or pods?

Reply:
As far as I understand, endpoint discovery is absolutely separate from pods/services, so it's up to the user what filters to apply. In a case when there are two blocks of discovery roles, like for endpoints and pods, then they will be discovered as separate targets.

Reply:
We could potentially miss the service/pod metadata though when doing endpoint discovery which would be unexpected for users I think.

Reply:
Can you please explain a bit a case when this situation is possible? 

As far as I understand, there are multiple roles for kubernetes discovery in Prometheus: node, service, pod, endpoints, ingress. Each of these roles maps 1-to-1 to a group of kubernetes resources. 

For endpoints, it queries the `/api/v1/namespaces/<ns>/endpoints/<name>` kubernetes API route. Endpoint is not a pod or service, it's a special type of internal intermediate resource used to map pods to services. The mapping is internal to kubernetes and not exposed to anywhere. So it can have it's own set of labels and fields we can filter with (however they are usually the same as for the corresponding pod).

In these particular lines of code we subscribe to all the updates for the `endpoints` route, `pods` and `services` are not touched anywhere here.

The only issue I see here is that user may have (for some reason) different set of selectors for pods and endpoints of the single service, and be surprised that the discovered IPs are different. But it seems quite logical for me: different filters => different discovered targets.

If you have an example when this filtering will not work or will confuse user, I would be really happy to take it into account and change the implementation.

Reply:
Read the rest of the case statement, just below this.

Reply:
Oops, got it, thanks for the hint. Will update the PR tomorrow to address this issue.